### ISSUE TITLE
[DOCS] `--prefix` not documented for `npm run`, `npm install`, etc.

### ISSUE THREAD

**Issue by devuxer** [npm/cli#1368]
Under [npm-run-scripts](https://docs.npmjs.com/cli/run-script), it says `npm run-script <command> [--silent] [-- <args>...]`, but it doesn't say anything about `--prefix`.

The only place I've seen so far that mentions `--prefix` is [shorthands and other CLI niceties](https://docs.npmjs.com/misc/config#shorthands-and-other-cli-niceties), but this is not likely to be seen, so many developers will not realize the feature exists.

I would propose something like this for npm-run-script:

----

**SYNOPSIS**

`npm run-script <command> [--silent]  [--if-present] [--prefix <path>] [--scripts-prepend-node-path | --scripts-prepend-node-path=auto] [-- <args>...]`

[...]

**Options:**

- `--silent` - prevents npm ERR! from showing in the output on error
- `--if-present` - avoids exiting with a non-zero exit code when the script is undefined, letting potentially undefined scripts to be run without breaking the execution chain
- `--prefix <path>` - forces non-global commands to run in the specified folder, letting scripts contained in other project folders be run without needing to switch to them first (as might occur with a monorepo)
- `--scripts-prepend-node-path` - add the directory within which node resides to the PATH
    - `--scripts-prepend-node-path=auto` - only adds the directory to the path if it is not already there (this has been the default since **npm v3**)

----

**Comment by darcyclarke** [comment]
npm `v6` is no longer in active development; We will continue to push security releases to `v6` at our team's discretion as-per our [Support Policy](https://github.com/npm/cli/wiki/Support-Policy).

If your bug is preproducible on `v7`, please re-file this issue using our [new issue template](https://github.com/npm/cli/issues/new?assignees=&labels=Bug%2CNeeds+Triage%2CRelease+7.x&template=bug_7.yml&title=%5BBUG%5D+%3Ctitle%3E).

If your issue was a feature request, please consider opening a new [RRFC](https://github.com/npm/rfcs/issues) or [RFC](https://github.com/npm/rfcs). If your issue was a question or other idea that was not CLI-specific, consider opening a discussion on our [feedback repo](https://github.com/npm/feedback)

**Closing:** This is an automated message.

**Comment by devuxer** [comment]
@darcyclarke , This was just a documentation issue that nobody noticed for a full year, so I'm not interesting in re-filing.  If you have the power to change the label to Release 7.x, though, that would be great.

**Comment by KishkinJ10** [comment]
@wraithgar Sir can I work on this ??

**Comment by wraithgar** [comment]
Prefix is one of those 'core' concepts that doesn't make a lot of sense to call out for every command, but may benefit from mentioning in the full docs.

Prefix is mentioned in the `npm install` docs for example, but briefly and only in relation to the `global` flag.

If you want to try adding examples to the commands where you think it makes sense start with the .md files that power those pages, for example: https://github.com/npm/cli/blob/latest/docs/content/commands/npm-install.md

**Comment by devuxer** [comment]
@wraithgar,

I respectfully disagree. The `prefix` flag has extremely poor discoverability in my experience, and it's not really a "core concept" if most people are unaware it exists. I could see an argument for putting an "other flags" link that points to a page with flags that are common to many commands, but the way the docs are now, it' tricks developers into thinking they are seeing *all* the flags when in fact, they are only see a subset.

**Comment by ljharb** [comment]
The prefix config can cause all sorts of problems; it’s best if it’s *not* discoverable.

**Comment by devuxer** [comment]
@ljharb, If that's the case, then why aren't the problems it can cause also documented?

**Comment by gondar00** [comment]
@devuxer can you please once explain the use of `--prefix` so that people who land up on this issue have a better idea

**Comment by rcrimp** [comment]
I stumbled upon this issue while looking for documentation on `--prefix`

I have a genuine use case for using `--prefix`

I am setting up a dev environment for education. Students do not have root privilege for security reasons. The student's home directories are mounted over a network, and because of that, `npm install x` always fails. We suspect it's the same root issue as [this](https://stackoverflow.com/questions/66168203/npm-install-code-enoent-syscall-rename-issue), where npm install fails if dropbox is actively syncing the current directory.

```bash
$ npm install http-server
npm ERR! code ENOENT
npm ERR! syscall rename
npm ERR! path /Users/crire682/.npm/_cacache/tmp/cc11f80d
npm ERR! dest /Users/crire682/.npm/_cacache/content-v2/sha512/22/8b/63a5ad166cb158b38d9d9510049240065560cbf0cf41e1b4404a4c339836c9d17d081c43ea8c3915aa151bb45ccf44a0f045547359f7d34b442b714e1a1f
npm ERR! errno ENOENT
npm ERR! enoent Invalid response body while trying to fetch https://registry.npmjs.org/lodash: ENOENT: no such file or directory, rename '/Users/crire682/.npm/_cacache/tmp/cc11f80d' -> '/Users/crire682/.npm/_cacache/content-v2/sha512/22/8b/63a5ad166cb158b38d9d9510049240065560cbf0cf41e1b4404a4c339836c9d17d081c43ea8c3915aa151bb45ccf44a0f045547359f7d34b442b714e1a1f'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/crire682/.npm/_logs/2022-06-21T00_48_50_256Z-debug-0.log
```

Our workaround is `npm install x --prefix .`
```bash
$ npm install http-server --prefix .

added 39 packages, and audited 40 packages in 6s

9 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```


It magically fixes everything. Clearly there is an underlying bug with npm when installing into a directory which is either cloud synced, or network mounted.

**Comment by ljharb** [comment]
@rcrimp an easier solution is to install node in each user's home directory, avoiding the need for escalated privileges (which is what https://nvm.sh does by default).

**Comment by rcrimp** [comment]
@ljharb thanks for the link to nvm, ill check it out.

We have installed node (and npm) globally. But our issue is when `npm install` saves tmp files in ~/.npm/ where ~ is on a network mounted file share.

(I'll make a separate issue if/when i have more details about reproducibility)

**Comment by ljharb** [comment]
Why wouldn’t every user have full permissions on their own home dir?

**Comment by pringshia** [comment]
> The prefix config can cause all sorts of problems; it’s best if it’s _not_ discoverable.

The Firebase docs mention the `--prefix` flag: https://firebase.google.com/docs/functions/typescript#using_an_existing_typescript_project

Even if not recommended, it would be nice if there was at least some documentation on it. A note could always be included mentioning not to use it.

**Comment by Yogu** [comment]
> The prefix config can cause all sorts of problems; it’s best if it’s not discoverable.

@ljharb could you give some pointers what kind of problems it can cause?

Also agree that the feature should be documented. The docs mention `{prefix}` in a few places, but I have not found any mention of `--prefix`.

**Comment by ljharb** [comment]
It breaks nvm, for one - for another, putting global modules in an arbitrary place encourages sharing global modules across node versions, which can't work reliably.

**Comment by Yogu** [comment]
Ah thanks, this makes sense.

My use case is running stuff like `npm install` or `npm publish` of several packages in a row in a CI script. All of the directories already have a `package.json`.

```
npm install --prefix lib
npm install --prefix app
npm publish --prefix lib
```

Could be done in a subshell, yes, but I think `--prefix` looks cleaner.

**Comment by ljharb** [comment]
¯\\\_(ツ)_/¯ a matter of opinion i guess; i think `(cd lib && npm install)` is much clearer since it relies on shell idioms instead of an obscure npm argument.

**Comment by wraithgar** [comment]
I think we're gonna close this as a `wontfix` as `--prefix` is not the best flag to be promoting.  We do eventually want to refactor or remove it but for now we should at least not be promoting it.

**Comment by Yogu** [comment]
Ok thank you, this at least clarifies that nobody should use it, and documentation suggesting it should be updated.

I added [an answer to the StackOverflow question](https://stackoverflow.com/a/73653667/880367) that prominently features this option and added a comment to the accepted answer that this feature is not officially supported.

**Comment by ArvinB** [comment]
I use the `--prefix` option a lot. Especially in `Dockerfile` for building images. I really hope this "hidden" feature will not be removed.

**Comment by EdanKriss** [comment]
> I use the `--prefix` option a lot. Especially in `Dockerfile` for building images. I really hope this "hidden" feature will not be removed.

I also rely on `--prefix` in my Dockerfiles.
I could live without this from a convenience point of view, however it remains the only solution I have found for a mysterious issue involving getting npm to use ssh credentials from a Docker BuildKit ssh mounted layer:
```
# npm only seems to use mounted ssh credentials from root directory, hence the --prefix.
RUN --mount=type=ssh,id=github npm ci --loglevel verbose  --prefix ${DEPLOY_PATH}
```
While I understand a convention favoring shell idioms, there are legitimate reasons to install modules to a directory other than the shell PWD, especially when you are not using the `-g` flag. 

Perhaps continue the roadmap of removing --prefix, and replace with a narrower, safer flag?

**Comment by BenJackGill** [comment]
For those who came here because `--prefix` was mentioned in [Firebase docs](https://firebase.google.com/docs/functions/typescript#using_an_existing_typescript_project) for `predeploy` and `postdeploy` hooks, I can confirm that the suggestion from @ljharb to use something like `cd [DIRECTORY] && npm run [SCRIPT]` also works well.

For example:

```JSON
{
  "functions": [
    {
      "source": "functions",
      "codebase": "default",
      "predeploy": [
        "cd \"$RESOURCE_DIR\" && npm run myCustomScript",
        "cd \"$RESOURCE_DIR\" && npm run lint",
        "cd \"$RESOURCE_DIR\" && npm run build"
      ],
      "postdeploy": [
        "cd \"$RESOURCE_DIR\" && npm run myCustomScript"
      ]
    }
  ],
}
```
